In [1]:
import pandas as pd
from matplotlib import pyplot as plt
import numpy as np

In [2]:
drives = pd.read_csv("../data/processed/drives.csv")
drives

game_id team_id  num  quarter   time  los  plays length  \
0      2009_1_TEN_PIT     PIT    1        1  15:00   58      3   1:44   
1      2009_1_TEN_PIT     TEN    1        1  13:16   98      3   1:52   
2      2009_1_TEN_PIT     PIT    2        1  11:24   43      5   3:04   
3      2009_1_TEN_PIT     TEN    2        1   8:20   89      6   1:36   
4      2009_1_TEN_PIT     PIT    3        1   6:44   73      3   1:55   
...               ...     ...  ...      ...    ...  ...    ...    ...   
92968   2023_22_SF_KC      SF   11        4   5:46   25      7   3:53   
92969   2023_22_SF_KC      KC   12        4   1:53   25     11   1:50   
92970   2023_22_SF_KC      SF   12        4   0:03   25      1   0:03   
92971   2023_22_SF_KC      SF   13        5  15:00   25     13   7:38   
92972   2023_22_SF_KC      KC   13        5   7:22   25     13   7:19   

       net_yds       result  ... Def_False_Start Def_Illegal_Shift  \
0            2         Punt  ...               0                 0   
1            2         Punt  ...               0                 0   
2            2         Punt  ...               0                 0   
3           70    Missed FG  ...               0                 0   
4           -6         Punt  ...               0                 0   
...        ...          ...  ...             ...               ...   
92968       40   Field Goal  ...               0                 0   
92969       64   Field Goal  ...               0                 0   
92970       -1  End of Game  ...               0                 0   
92971       66   Field Goal  ...               0                 0   
92972       75    Touchdown  ...               0                 0   

       Def_Face_Mask  Off_Face_Mask  \
0                  0              0   
1                  0              0   
2                  0              0   
3                  0              0   
4                  0              0   
...              ...            ...   
92968              0              0   
92969              0              0   
92970              0              0   
92971              0              0   
92972              0              0   

       Def_Lowering_the_Head_to_Make_Forcible_Contact  \
0                                                   0   
1                                                   0   
2                                                   0   
3                                                   0   
4                                                   0   
...                                               ...   
92968                                               0   
92969                                               0   
92970                                               0   
92971                                               0   
92972                                               0   

       Off_Illegal_Kick/Kicking_Loose_Ball  total_off_pen  total_def_pen  \
0                                        0              0              0   
1                                        0              0              0   
2                                        0              0              0   
3                                        0              0              1   
4                                        0              0              0   
...                                    ...            ...            ...   
92968                                    0              0              0   
92969                                    0              0              0   
92970                                    0              0              0   
92971                                    0              0              0   
92972                                    0              2              1   

       total_off_pen_yards  total_def_pen_yards  
0                        0                    0  
1                        0                    0  
2                        0                    0  
3                        0   

While looking at the other drive results might be interesting, I'm going to focus on the most common results.

In [3]:
drives = drives[drives.result.isin(["Punt", "Touchdown", "Field Goal", "Interception", "Fumble", "Safety"])]
drives

game_id team_id  num  quarter   time  los  plays length  \
0      2009_1_TEN_PIT     PIT    1        1  15:00   58      3   1:44   
1      2009_1_TEN_PIT     TEN    1        1  13:16   98      3   1:52   
2      2009_1_TEN_PIT     PIT    2        1  11:24   43      5   3:04   
4      2009_1_TEN_PIT     PIT    3        1   6:44   73      3   1:55   
5      2009_1_TEN_PIT     TEN    3        1   4:49   74      6   3:11   
...               ...     ...  ...      ...    ...  ...    ...    ...   
92967   2023_22_SF_KC      KC   11        4  11:22   25     12   5:36   
92968   2023_22_SF_KC      SF   11        4   5:46   25      7   3:53   
92969   2023_22_SF_KC      KC   12        4   1:53   25     11   1:50   
92971   2023_22_SF_KC      SF   13        5  15:00   25     13   7:38   
92972   2023_22_SF_KC      KC   13        5   7:22   25     13   7:19   

       net_yds        result  ... Def_False_Start Def_Illegal_Shift  \
0            2          Punt  ...               0                 0   
1            2          Punt  ...               0                 0   
2            2          Punt  ...               0                 0   
4           -6          Punt  ...               0                 0   
5           30  Interception  ...               0                 0   
...        ...           ...  ...             ...               ...   
92967       69    Field Goal  ...               0                 0   
92968       40    Field Goal  ...               0                 0   
92969       64    Field Goal  ...               0                 0   
92971       66    Field Goal  ...               0                 0   
92972       75     Touchdown  ...               0                 0   

       Def_Face_Mask  Off_Face_Mask  \
0                  0              0   
1                  0              0   
2                  0              0   
4                  0              0   
5                  0              0   
...              ...            ...   
92967              0              0   
92968              0              0   
92969              0              0   
92971              0              0   
92972              0              0   

       Def_Lowering_the_Head_to_Make_Forcible_Contact  \
0                                                   0   
1                                                   0   
2                                                   0   
4                                                   0   
5                                                   0   
...                                               ...   
92967                                               0   
92968                                               0   
92969                                               0   
92971                                               0   
92972                                               0   

       Off_Illegal_Kick/Kicking_Loose_Ball  total_off_pen  total_def_pen  \
0                                        0              0              0   
1                                        0              0              0   
2                                        0              0              0   
4                                        0              0              0   
5                                        0              1              0   
...                                    ...            ...            ...   
92967                                    0              0              0   
92968                                    0              0              0   
92969                                    0              0              0   
92971                                    0              0              0   
92972                                    0              2              1   

       total_off_pen_yards  total_def_pen_yards  
0                        0                    0  
1                        0                    0  
2                        0                    0  
4                

In [4]:
"""
Add "net_penalty_yards" column, which is a quantitative measure of how much penalties benefitted an offense or a
defense in a specific drive. Positive values mean that the offense gained net penalty yards.
"""

drives["net_penalty_yards"] = drives.total_def_pen_yards - drives.total_off_pen_yards

drives

/tmp/ipykernel_2490829/880823554.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  drives["net_penalty_yards"] = drives.total_def_pen_yards - drives.total_off_pen_yards


game_id team_id  num  quarter   time  los  plays length  \
0      2009_1_TEN_PIT     PIT    1        1  15:00   58      3   1:44   
1      2009_1_TEN_PIT     TEN    1        1  13:16   98      3   1:52   
2      2009_1_TEN_PIT     PIT    2        1  11:24   43      5   3:04   
4      2009_1_TEN_PIT     PIT    3        1   6:44   73      3   1:55   
5      2009_1_TEN_PIT     TEN    3        1   4:49   74      6   3:11   
...               ...     ...  ...      ...    ...  ...    ...    ...   
92967   2023_22_SF_KC      KC   11        4  11:22   25     12   5:36   
92968   2023_22_SF_KC      SF   11        4   5:46   25      7   3:53   
92969   2023_22_SF_KC      KC   12        4   1:53   25     11   1:50   
92971   2023_22_SF_KC      SF   13        5  15:00   25     13   7:38   
92972   2023_22_SF_KC      KC   13        5   7:22   25     13   7:19   

       net_yds        result  ... Def_Illegal_Shift Def_Face_Mask  \
0            2          Punt  ...                 0             0   
1            2          Punt  ...                 0             0   
2            2          Punt  ...                 0             0   
4           -6          Punt  ...                 0             0   
5           30  Interception  ...                 0             0   
...        ...           ...  ...               ...           ...   
92967       69    Field Goal  ...                 0             0   
92968       40    Field Goal  ...                 0             0   
92969       64    Field Goal  ...                 0             0   
92971       66    Field Goal  ...                 0             0   
92972       75     Touchdown  ...                 0             0   

       Off_Face_Mask  Def_Lowering_the_Head_to_Make_Forcible_Contact  \
0                  0                                               0   
1                  0                                               0   
2                  0                                               0   
4                  0                                               0   
5                  0                                               0   
...              ...                                             ...   
92967              0                                               0   
92968              0                                               0   
92969              0                                               0   
92971              0                                               0   
92972              0                                               0   

       Off_Illegal_Kick/Kicking_Loose_Ball  total_off_pen  total_def_pen  \
0                                        0              0              0   
1                                        0              0              0   
2                                        0              0              0   
4                                        0              0              0   
5                                        0              1              0   
...                                    ...            ...            ...   
92967                                    0              0              0   
92968                                    0              0              0   
92969                                    0              0              0   
92971                                    0              0              0   
92972                                    0              2              1   

       total_off_pen_yards  total_def_pen_yards  net_penalty_yards  
0                        0                    0                  0  
1                        0                    0                  0  
2                        0                    0                  0  
4                        0                    0                  0  
5                        5                    0                 -5  
...                    ...                  ...                ...  
92967                    0                    0

In [5]:
drives.groupby("result").net_penalty_yards.mean()

result
Field Goal      0.645262
Fumble          0.120981
Interception    0.265965
Punt           -1.330627
Safety         -2.376147
Touchdown       2.136246
Name: net_penalty_yards, dtype: float64

On the average drive that results in a touchdown, the offense **gains** an average of 2.14 yards due to penalties. On the average drive that results in a field goal, the offense **gains** an average of 0.65 yards due to penalties. In contrast, on the average drive that end in a safety, the offense **loses** an average of 2.37 yards due to penalties, and on the average drive that ends in a punt, the offense **loses** an average of 1.33 yards due to penalties. 

So, it is clear that if we look at drives that result in offensive scores versus drives that don't, it is clear that drives that have more points scored tend to have penalties that benefit the offense more than the defense.

To dig deeper, lets look at which penalties tend to have the biggest impact on drives.

In [6]:
from sklearn.feature_selection import chi2, SelectKBest

# penalty columns
X = drives.iloc[:,12:94]

# drive results
y = drives.result

k_best_selector = SelectKBest(score_func=chi2)
k_best_selector.fit(X, y)

feature_score_pairs = zip(X.columns, k_best_selector.scores_)
sorted_feature_score_pairs = sorted(zip(X.columns, k_best_selector.scores_), key=lambda x: x[1], reverse=True)

penalties_by_importance = pd.DataFrame({"Penalties": X.columns, "Importance": k_best_selector.scores_})
penalties_by_importance = penalties_by_importance.sort_values(by='Importance', ascending=False)
penalties_by_importance.head(10)

Penalties   Importance
3               Def_Pass_Interference  2703.755949
21                        Def_Holding  1442.927989
13                        Def_Offside   907.806992
0           Def_Unnecessary_Roughness   883.301925
19            Def_Roughing_the_Passer   739.660113
23          Off_Intentional_Grounding   668.872418
38  Def_Illegal_Block_Above_the_Waist   608.256425
4            Def_Face_Mask_(15_Yards)   392.917101
8                 Def_Illegal_Contact   372.955352
11           Def_Illegal_Use_of_Hands   343.049531

Here, we take a look at which penalties tend to have the largest affect on the outcome of a drive. `sklearn.feature_selection.SelectKBest` gives features an importance score based on a univariate statistical test. In this case, we are using the chi-squared test as a scoring function.

We can see that Defensive Pass Interference has by far the most importance in determing the outcome of a drive. This makes sense because Defense Pass Interference is a spot foul, which means that the penalty can potentially cost the defense an unlimited amount of yardage.

However, we can also see that almost all of the 10 most important penalties in determining the outcome of a drive are defensive penalties! This is definitely not what I would have expected to see. The one exception is Intentional Grounding. This makes sense because intentional grounding is somewhat likely to lead to safeties, and otherwise has a major impact on an offense because it is a spot foul and a loss of down.

Lets take a look at the next 10 most important penalties:

In [7]:
penalties_by_importance.iloc[10:20,:]

Penalties  Importance
22  Def_Neutral_Zone_Infraction  282.445376
2                   Off_Holding  264.966547
7               Off_False_Start  238.927451
56      Off_Horse_Collar_Tackle  188.477379
5              Def_Encroachment  175.973369
17  Def_Unsportsmanlike_Conduct  171.842131
71    Def_Too_Many_Men_on_Field  160.595766
34              Def_12_On-field  146.896185
69  Def_Illegal_Blindside_Block  146.030641
40      Def_Horse_Collar_Tackle  144.390932

Offensive holding appears to be the next most impactful offensive penalty. This makes sense because offensive holding is a fairly common penalty that sets the offense back 10 yards, which can have a major impact on slowing down drives.

TODO: We could try to model the "expected points added" of specific types of penalties.